hw/dv/verilator/cpp/scrambled_ecc32_mem_area.cc

```diff
@@ -0,0 +1,164 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "scrambled_ecc32_mem_area.h"
+#include "scramble_model.h"
+
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+#include <sstream>
+
+#include "sv_scoped.h"
+
+// This is the maximum width of a nonce that's supported by the code in
+// prim_util_get_scramble_key_nonce.svh
+static const uint32_t kScrMaxNonceWidth = 320;
+static const uint32_t kScrMaxNonceWidthBytes = (kScrMaxNonceWidth + 7) / 8;
+
+// Functions to convert from integer address to/from a vector of bytes,
+// addr_width is given in bits
+static std::vector<uint8_t> AddrIntToBytes(uint32_t addr, uint32_t addr_width) {
+  uint32_t addr_width_bytes = (addr_width + 7) / 8;
+  std::vector<uint8_t> addr_bytes(addr_width_bytes);
+
+  for (int i = 0; i < addr_width_bytes; ++i) {
+    addr_bytes[i] = addr & 0xff;
+    addr >>= 8;
+  }
+
+  return addr_bytes;
+}
+
+static uint32_t AddrBytesToInt(const std::vector<uint8_t> &addr) {
+  uint32_t addr_out = 0;
+  int cur_shift = 0;
+
+  for (int i = 0; i < addr.size(); ++i) {
+    addr_out |= (addr[i] << cur_shift);
+    cur_shift += 8;
+  }
+
+  return addr_out;
+}
+
+extern "C" {
+int simutil_get_scramble_key(svBitVecVal *key);
+int simutil_get_scramble_nonce(svBitVecVal *nonce);
+}
+
+std::vector<uint8_t> ScrambledEcc32MemArea::GetScrambleKey() const {
+  SVScoped scoped(scr_scope_);
+  uint8_t key_minibuf[kPrinceWidthByte * 2];
+
+  if (!simutil_get_scramble_key((svBitVecVal *)key_minibuf)) {
+    std::ostringstream oss;
+    oss << "Could not read key at scope " << scr_scope_;
+    throw std::runtime_error(oss.str());
+  }
+
+  return std::vector<uint8_t>(&key_minibuf[0],
+                              &key_minibuf[kPrinceWidthByte * 2]);
```

Comment:
This is really neat, but maybe it would be best to do the casting with the simutil_get_scramble_key call explicitly?

We're casting between `svBitVecVal*` and `unsigned char*`, so I don't think this is technically UB, but it definitely does depend on implementation-defined endianness (because an `svBitVecVal` is essentially a `uint32_t`). I'd probably suggest an explicit loop that extracts 8 bits at a time with a shift. (Maybe the `AddrIntToBytes` function above could use it too).

Reply:
I've introduced a separate `ByteVecFromSV` function. I didn't use it for `AddrIntToBytes` as it's not working with arrays of uint32_t just the one so it didn't fit nicely.

Though as svBitVecVal is defined as `uint32_t` in the SV spec (I think... not entirely sure if the header given in the spec that defines it is an example or meant to be part of the specification)  I could make it take a `uint32_t val[]` and then pass a pointer to the address passed to `AddrIntToBytes`.